File: plugins/virtualfs/skills/py-sdk/examples/ingest-explore.py
"""Bulk-ingest a local folder, then explore the codebase via exec_batch.

Demonstrates the canonical agent workflow:
  1. ONE round-trip via `sb.ingest_files(...)` to bootstrap the sandbox.
  2. ALL subsequent reads/lists via `sb.exec_batch([...])` (one round-trip).
  3. Clean up.

Compare wall-clock cost: per-file `fs.write` would cost N HTTP round-trips
(seconds-per-file at typical RTT). `ingest_files` is one round-trip regardless
of file count.

Run with:
    BASE_URL=... AUTH_SECRET=... python ingest-explore.py /path/to/local/code
"""

from __future__ import annotations

import os
import sys
import time
from pathlib import Path

from virtualfs import Client

SANDBOX_BASE = "/home/user/proj"
SKIP_DIRS = {"__pycache__", ".git", ".mypy_cache", ".pytest_cache", "node_modules"}
SKIP_SUFFIXES = {".pyc", ".pyo"}


def collect(root: Path) -> dict[str, bytes]:
    out: dict[str, bytes] = {}
    for p in root.rglob("*"):
        if any(part in SKIP_DIRS for part in p.parts):
            continue
        if p.suffix in SKIP_SUFFIXES or p.name == ".DS_Store":
            continue
        if p.is_file():
            out[p.relative_to(root).as_posix()] = p.read_bytes()
    return out


def main() -> int:
    if len(sys.argv) != 2:
        print(f"usage: {sys.argv[0]} <local-folder>", file=sys.stderr)
        return 2
    folder = Path(sys.argv[1]).expanduser().resolve()
    if not folder.is_dir():
        print(f"not a directory: {folder}", file=sys.stderr)
        return 2

    print(f"walking {folder} ...")
    files = collect(folder)
    total = sum(len(v) for v in files.values())
    print(f"  {len(files)} files, {total:,} bytes")

    with Client(
        base_url=os.environ["BASE_URL"],
        auth_secret=os.environ["AUTH_SECRET"],
        sub="ingest-explore",
    ) as fs:
        sb = fs.sandboxes.create(name="ingest-explore")
        try:
            t = time.perf_counter()
            sb.ingest_files(files, base_path=SANDBOX_BASE)
            print(f"ingest_files: {(time.perf_counter() - t) * 1000:.0f} ms "
                  f"({len(files)} files, 1 HTTP round-trip)")

            # Build an exploration probe-set. Every probe runs in ONE batch
            # round-trip. Add or trim freely — the batch endpoint handles up to
            # 50 scripts and shares a single timeout budget.
            probes = [
                ("tree",
                 f"find {SANDBOX_BASE} -type f -printf '%s %p\\n' | sort -rn | head -10"),
                ("py_count",
                 f"find {SANDBOX_BASE} -name '*.py' | wc -l"),
                ("imports",
                 f"grep -rhn '^import \\|^from ' {SANDBOX_BASE} 2>/dev/null "
                 f"| sort -u | head -20"),
                ("classes",
                 f"grep -rhn '^class ' {SANDBOX_BASE} 2>/dev/null | head -15"),
                ("todos",
                 f"grep -rn 'TODO\\|FIXME' {SANDBOX_BASE} 2>/dev/null | head -10"),
                ("entry_point",
                 f"find {SANDBOX_BASE} -maxdepth 2 -name '__init__.py' -o "
                 f"-name 'main.py' -o -name 'app.py' | head -5"),
            ]

            t = time.perf_counter()
            results = sb.exec_batch(
                [{"id": label, "script": script} for label, script in probes],
                timeout_ms=60_000,
            )
            print(f"exec_batch:    {(time.perf_counter() - t) * 1000:.0f} ms "
                  f"for {len(probes)} probes (1 round-trip)")

            for r in results:
                head = r.stdout.rstrip().splitlines()[:8]
                print(f"\n[{r.id}] exit={r.exit_code}")
                for line in head:
                    print(f"  {line}")
        finally:
            fs.sandboxes.delete(sb.id)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: plugins/sql-fs/skills/ts-sdk/examples/ingest-explore.ts
/**
 * Bulk-ingest a local folder, then explore the codebase via execBatch.
 *
 * Demonstrates the canonical agent workflow:
 *   1. ONE round-trip via `sb.ingestFiles(...)` to bootstrap the sandbox.
 *   2. ALL subsequent reads/lists via `sb.execBatch([...])` (one round-trip).
 *   3. Clean up.
 *
 * Compare wall-clock cost: per-file `fs.write` would cost N HTTP round-trips
 * (seconds-per-file at typical RTT). `ingestFiles` is one round-trip regardless
 * of file count.
 *
 * Run with:
 *   BASE_URL=... AUTH_SECRET=... npx tsx ingest-explore.ts /path/to/local/code
 */

import { readFile, readdir } from "node:fs/promises";
import path from "node:path";
import { Client } from "sql-fs-sdk";

const SANDBOX_BASE = "/home/user/proj";
const SKIP_DIRS = new Set([".git", "node_modules", "dist", ".next", "__pycache__"]);
const SKIP_NAMES = new Set([".DS_Store"]);

async function collect(root: string, base = root, out: Record<string, Uint8Array> = {}): Promise<Record<string, Uint8Array>> {
	for (const entry of await readdir(root, { withFileTypes: true })) {
		if (entry.isDirectory()) {
			if (SKIP_DIRS.has(entry.name)) continue;
			await collect(path.join(root, entry.name), base, out);
		} else if (entry.isFile() && !SKIP_NAMES.has(entry.name)) {
			const abs = path.join(root, entry.name);
			const rel = path.relative(base, abs).split(path.sep).join("/");
			out[rel] = await readFile(abs);
		}
	}
	return out;
}

async function main(): Promise<number> {
	const folder = process.argv[2];
	if (!folder) {
		console.error(`usage: ${process.argv[1]} <local-folder>`);
		return 2;
	}

	console.log(`walking ${folder} ...`);
	const files = await collect(path.resolve(folder));
	const total = Object.values(files).reduce((n, b) => n + b.byteLength, 0);
	console.log(`  ${Object.keys(files).length} files, ${total.toLocaleString()} bytes`);

	const client = new Client({
		baseUrl: process.env.BASE_URL!,
		authSecret: process.env.AUTH_SECRET!,
		sub: "ingest-explore",
	});
	try {
		const sb = await client.sandboxes.create({ name: "ingest-explore" });
		try {
			let t = performance.now();
			await sb.ingestFiles(files, { basePath: SANDBOX_BASE });
			console.log(
				`ingestFiles: ${Math.round(performance.now() - t)} ms ` +
					`(${Object.keys(files).length} files, 1 HTTP round-trip)`,
			);

			// Build an exploration probe-set. Every probe runs in ONE batch
			// round-trip. Add or trim freely — the batch endpoint handles up to
			// 50 scripts and shares a single timeout budget.
			const probes: Array<{ id: string; script: string }> = [
				{ id: "tree", script: `find ${SANDBOX_BASE} -type f -printf '%s %p\\n' | sort -rn | head -10` },
				{ id: "ts_count", script: `find ${SANDBOX_BASE} -name '*.ts' | wc -l` },
				{ id: "imports", script: `grep -rhn '^import ' ${SANDBOX_BASE} 2>/dev/null | sort -u | head -20` },
				{ id: "exports", script: `grep -rhn '^export ' ${SANDBOX_BASE} 2>/dev/null | head -15` },
				{ id: "todos", script: `grep -rn 'TODO\\|FIXME' ${SANDBOX_BASE} 2>/dev/null | head -10` },
				{ id: "entry", script: `find ${SANDBOX_BASE} -maxdepth 2 -name 'index.ts' -o -name 'main.ts' | head -5` },
			];

			t = performance.now();
			const results = await sb.execBatch(probes, { timeoutMs: 60_000, readOnly: true });
			console.log(`execBatch:   ${Math.round(performance.now() - t)} ms for ${probes.length} probes (1 round-trip)`);

			for (const r of results) {
				const head = r.stdout.trimEnd().split("\n").slice(0, 8);
				console.log(`\n[${r.id}] exit=${r.exitCode}`);
				for (const line of head) console.log(`  ${line}`);
			}
		} finally {
			await client.sandboxes.delete(sb.id);
		}
	} finally {
		client.close();
	}
	return 0;
}

process.exit(await main());
